src/a_star.py

```diff
@@ -39,13 +40,244 @@ def return_path(current_node):
+            if stop == 0:
+                break
+
+        if i < len(end_euclidean_distances):
+            destiny = end_euclidean_distances[i][0]
+            paths, borders = astar_2d(maze[end[0]], end_2d, destiny, allow_diagonal_movement)
+            paths = add_floor_to_steps(paths, end[0])
+            borders = add_floor_to_steps(borders, end[0])
+            path_step += paths
+            border_step += borders
+            path = paths[-1]
+            if path is not None:
+                start_calculated_destinations = [x[0] for x in paths_in_start_floor]
+                if destiny in start_calculated_destinations:
+                    start_path = paths_in_start_floor[start_calculated_destinations.index(destiny)][1]
+                    path.reverse()
+                    path_step.append(start_path + path)
+                    border_step.append(None)
+                    return path_step, border_step
+                paths_in_end_floor.append((destiny, path))
+        else:
+            stop -= 1
+            if stop == 0:
+                break
+
+        i += 1
+    return None
+
+def astar_2d(
+    maze: np.ndarray,
+    start: Tuple[int, int],
+    end: Tuple[int, int],
+    allow_diagonal_movement: bool = False
+) -> Tuple[List[List[Tuple[int, int]]], List[List[Tuple[int, int]]]]:
+    """
+    Returns a list of tuples as a path from the given start to the given end in the given maze
+    :param maze:
+    :param start:
+    :param end:
+    :return:
+    """
+
+    #list of lists with the path at each step
+    path_step = []
+    border_step = []
+
+    # Create start and end node
+    start_node = Node(None, start)
+    start_node.g = start_node.h = start_node.f = 0
+    end_node = Node(None, end)
+    end_node.g = end_node.h = end_node.f = 0
+
+    # Initialize both open and closed list
+    open_list = []
+    closed_list = []
+
+    # Heapify the open_list and Add the start node
+    heapq.heapify(open_list) 
+    heapq.heappush(open_list, start_node)
+
+    # Adding a stop condition
+    outer_iterations = 0
+    max_iterations = (len(maze[0]) * len(maze) // 2)
+
+    diagonal_cost = 5.0
+    horizontal_cost = 1.0
+
+    adjacent_squares = ((0, -1), (0, 1), (-1, 0), (1, 0))
+    direction_cost= (horizontal_cost, horizontal_cost, horizontal_cost, horizontal_cost)
+    adjacent_square_pick_index = [0, 1, 2, 3]
+
+    # what squares do we search
+    if allow_diagonal_movement:
+        adjacent_squares = ((0, -1), (0, 1), (-1, 0), (1, 0), (-1, -1), (-1, 1), (1, -1), (1, 1))
+        direction_cost = (horizontal_cost, horizontal_cost, horizontal_cost, horizontal_cost, diagonal_cost, diagonal_cost, diagonal_cost, diagonal_cost)
+        adjacent_square_pick_index = [0, 1, 2, 3, 4, 5, 6, 7]
+
+    # Loop until you find the end
+    while len(open_list) > 0:
+        #Randomize the order of the adjacent_squares_pick_index to avoid a decision making bias
+        random.shuffle(adjacent_square_pick_index)
```

Comment:
I don't understand why this random call is necessary, why would favouring any particular direction be a problem? The cost dictates the best path anyways

Reply:
it's not a good idea, this is useless